component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/voiceotp/VoiceOTPAuthenticator.java

```diff
@@ -0,0 +1,2034 @@
+        boolean isLocalUser = VoiceOTPUtils.isLocalUser(context);
+
+        if (authenticatedUser != null && isLocalUser && VoiceOTPUtils.isAccountLocked(authenticatedUser)) {
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Authentication failed since authenticated user: %s,  account is locked.",
+                        authenticatedUser));
+            }
+            context.setProperty(VoiceOTPConstants.ACCOUNT_LOCKED, true);
+            throw new AuthenticationFailedException("User account is locked.");
+        }
+
+        String userToken = request.getParameter(VoiceOTPConstants.CODE);
+        String contextToken = (String) context.getProperty(VoiceOTPConstants.OTP_TOKEN);
+        if (StringUtils.isEmpty(request.getParameter(VoiceOTPConstants.CODE))) {
+            throw new InvalidCredentialsException("Code cannot not be null");
+        }
+        if (Boolean.parseBoolean(request.getParameter(VoiceOTPConstants.RESEND))) {
+
+            log.debug("Retrying to resend the OTP");
+
+            throw new InvalidCredentialsException("Retrying to resend the OTP");
+        }
+
+        if (context.getProperty(VoiceOTPConstants.MOBILE_NUMBER_UPDATE_FAILURE) != null) {
+            context.setProperty(VoiceOTPConstants.MOBILE_NUMBER_UPDATE_FAILURE, "false");
+        }
+
+        boolean succeededAttempt = false;
+        if (userToken.equals(contextToken)) {
+            context.removeProperty(VoiceOTPConstants.CODE_MISMATCH);
+            processValidUserToken(context, authenticatedUser);
+            succeededAttempt = true;
+        } else if (isLocalUser && "true".equals(VoiceOTPUtils.getBackupCode(context))) {
+            succeededAttempt = checkWithBackUpCodes(context, userToken, authenticatedUser);
+        } else {
+
+            log.debug("Given otp code is a mismatch.");
+
+            context.setProperty(VoiceOTPConstants.CODE_MISMATCH, true);
+        }
+
+        if (succeededAttempt && isLocalUser) {
+            String username = String.valueOf(context.getProperty(VoiceOTPConstants.USER_NAME));
+            String mobileNumber;
+            try {
+                mobileNumber = VoiceOTPUtils.getMobileNumberForUsername(username);
+            } catch (VoiceOTPException e) {
+                throw new AuthenticationFailedException("Failed to get the parameters from authentication xml file " +
+                        "for user:  " + username + " for tenant: " + context.getTenantDomain(), e);
+            }
+
+            if (StringUtils.isBlank(mobileNumber)) {
+                String tenantDomain = MultitenantUtils.getTenantDomain(username);
+                Object verifiedMobileObject = context.getProperty(VoiceOTPConstants.REQUESTED_USER_MOBILE);
+                if (verifiedMobileObject != null) {
+                    try {
+                        updateMobileNumberForUsername(context, request, username, tenantDomain);
```

Comment:
Are we allowing to update the mobile number if not available by default? or is there a config for that?

Reply:
we have a config for this(CaptureAndUpdateMobileNumber). This will determine pop up for enter the mobile number if mobile number is not configured already. If this is enabled and user entered the correct otp , the mobile number will be updated.

Reply:
Don't we need to check that config here?

Reply:
No need, this will only happen the current mobile number is blank for the user. If current mobile number is blank that means there is no other way to come to this point without prompting to enter the mobile number . Which means the config check is there